src/wallet/spend.rs

```diff
@@ -0,0 +1,403 @@
+    pub fn spend_from_wallet(
+        &mut self,
+        feerate: f64,
+        destination: Destination,
+        coins_to_spend: &[(ListUnspentResultEntry, UTXOSpendInfo)],
+    ) -> Result<Transaction, WalletError> {
+        log::info!("Creating Direct-Spend from Wallet.");
+
+        let mut coins = Vec::<(ListUnspentResultEntry, UTXOSpendInfo)>::new();
+
+        for coin in coins_to_spend {
+            // filter all contract and fidelity utxos.
+            if let UTXOSpendInfo::FidelityBondCoin { .. }
+            | UTXOSpendInfo::HashlockContract { .. }
+            | UTXOSpendInfo::TimelockContract { .. } = coin.1
+            {
+                log::warn!("Skipping Fidelity Bond or Contract UTXO.");
+                continue;
+            } else {
+                coins.push(coin.to_owned());
+            }
+        }
+
+        let tx = self.spend_coins(&coins, destination, feerate)?;
+
+        Ok(tx)
+    }
+
+    /// Redeem a Fidelity Bond.
+    /// This functions creates a spending transaction from the fidelity bond, signs and broadcasts it.
+    /// Returns the txid of the spending tx, and mark the bond as spent.
+    pub fn redeem_fidelity(&mut self, idx: u32, feerate: f64) -> Result<Txid, WalletError> {
+        let (bond, _, is_spent) = self
+            .store
+            .fidelity_bond
+            .get(&idx)
+            .ok_or(FidelityError::BondDoesNotExist)?;
+
+        if *is_spent {
+            return Err(FidelityError::BondAlreadySpent.into());
+        }
+        let utxo_spend_info = UTXOSpendInfo::FidelityBondCoin {
+            index: idx,
+            input_value: bond.amount,
+        };
+        let change_addr = &self.get_next_internal_addresses(1)?[0];
+        let destination = Destination::Sweep(change_addr.clone());
+        let all_utxo = self.list_fidelity_spend_info(None)?;
+        let mut utxo: Option<ListUnspentResultEntry> = None;
+        for (utxo_data, spend_info) in all_utxo {
+            if let UTXOSpendInfo::FidelityBondCoin { index, input_value } = spend_info.clone() {
+                if index == idx && input_value == bond.amount {
+                    utxo = Some(utxo_data)
+                }
+            }
+        }
+        let utxo = utxo.ok_or(FidelityError::BondAlreadySpent)?;
+
+        let tx = self.spend_coins(&vec![(utxo, utxo_spend_info)], destination, feerate)?;
+
+        let txid = self.send_tx(&tx)?;
+
+        log::info!("Fidelity redeem transaction broadcasted. txid: {}", txid);
+
+        // No need to wait for confirmation as that will delay the rpc call. Just send back the txid.
+
+        // mark is_spent
+        {
+            let (_, _, is_spent) = self
+                .store
+                .fidelity_bond
+                .get_mut(&idx)
+                .ok_or(FidelityError::BondDoesNotExist)?;
+
+            *is_spent = true;
+        }
+
+        Ok(txid)
+    }
+
+    #[allow(unused)]
+    pub fn spend_coins(
```

Comment:
make this pub(crate). If user isn't careful  they can mix different utxo types in a spend and deanonimize the swaps.. Better to make it private. All API user should instead use the wrapper APIs.